waspc/waspls/src/Wasp/LSP/SignatureHelp.hs

```diff
@@ -0,0 +1,261 @@
+  (MonadState ServerState m, MonadLog m) =>
+  LSP.Position ->
+  m LSP.SignatureHelp
+getSignatureHelpAtPosition position = do
+  src <- gets (^. currentWaspSource)
+  gets (^. cst) >>= \case
+    Nothing ->
+      -- No CST in the server state, can't create a signature.
+      return emptyHelp
+    Just syntax -> do
+      let offset = lspPositionToOffset src position
+      let location = locationAtOffset offset (fromSyntaxForest syntax)
+      findSignatureAtLocation src location >>= \case
+        Nothing -> do
+          logM "[getSignatureHelpAtPosition] no signature found"
+          return emptyHelp
+        Just signature -> do
+          logM "[getSignatureHelpAtPosition] found a signature"
+          let signatureFragments = signatureToFragments signature
+          let params = getParamsFromFragments signatureFragments
+          let activeParam = signatureParam signature >>= findActiveParameterIndex signatureFragments
+          logM $ "[getSignatureHelpAtPosition] at param idx = " ++ show activeParam ++ " params = " ++ show params
+          return $
+            LSP.SignatureHelp
+              { _signatures =
+                  LSP.List
+                    [ LSP.SignatureInformation
+                        { _parameters = Just $ LSP.List params,
+                          -- NOTE: VSCode higlights the 0th parameter if
+                          -- activeParameter is Nothing, so it's set to -1 when
+                          -- we don't want any param highlighted.
+                          _activeParameter = Just $ fromMaybe (-1) activeParam,
+                          _label = Text.pack $ showFragments signatureFragments,
+                          _documentation = Nothing
+                        }
+                    ],
+                _activeSignature = Just 0,
+                _activeParameter = Nothing
+              }
+  where
+    emptyHelp =
+      LSP.SignatureHelp
+        { _signatures = LSP.List [],
+          _activeSignature = Nothing,
+          _activeParameter = Nothing
+        }
+
+-- | 'Signature' describes the expected type at a specific location inside of
+-- a container in a wasp file. Every signature includes the type of the container
+-- surrounding the location. For example, the container type for a cursor
+-- positioned inside of a dictionary would be the type of the dictionary.
+--
+-- When the location is at a more specific place inside of the container--for
+-- example, immediately after @key: @ in a dictionary--then we say the location
+-- is at a parameter, and the parameter is described in relation to the container
+-- with an 'ExprKey'.
+data Signature
+  = -- | Inside a container, but not at any parameter inside of it.
+    InContainer !Type
+  | -- | Inside a container at the parameter specified by the 'ExprKey'. This
+    -- is named \"param\" to be consistent with the naming conventions used by
+    -- the LSP.
+    AtParam !Type !ExprKey
+  deriving (Eq, Show)
```

Comment:
This might be a bit nicer:

```suggestion
data Signature
  = -- | Type describes container, while ExprKey desribes param inside of container.
    InContainerAtParam !Type !ExprKey
  | InContainer !Type
  deriving (Eq, Show)
```

Although, would it make more sense to go with:
```suggestion
data Signature = Signature {
  _container :: !Type,
  _param :: !(Maybe ExprKey)
}
```

Then I don't think you even need any comments, almost.